### PR TITLE
Work with basic-auth enabled SolrCloud clusters

### LINKED ISSUE #218: Enhancements for working with basic auth enabled Solr clusters
For background on Solr basic auth, see: 
https://lucene.apache.org/solr/guide/8_8/basic-authentication-plugin.html#basic-authentication-plugin

The Solr operator should provide an easy way to enable basic authentication for a SolrCloud cluster.

There are two main uses cases we should support:

1. The operator should bootstrap a default `security.json` with basic auth enabled. This allows users to easily spin-up a secure Solr cluster without having to know all the details of how Solr security works. Having the Solr API be protected by basic-auth out of the box leads to a better getting started experience for Solr and the operator in general.
2. The operator should allow users to supply their own `security.json` (via ConfigMap) and corresponding secret with credentials the operator should use to make API requests to Solr. Both are needed and if the user chooses to supply their own `security.json` then they need to make sure the credentials are in the secret as well. For this use case, the operator's main job is to bootstrap the `security.json` in ZK as part of the initialization process.

When basic auth is enabled, we need to ensure the liveness / readiness probes work (`/admin/info/system` endpoint). In addition, the rolling restart logic requires access to the `/admin/collections` endpoint. Also, backup / restore requires API access. Thus, the Solr operator needs access to a username and password to make API calls. These credentials will come from a Kubernetes secret in the same namespace.

For the first use case (bootstrap a default config), the only tricky part is we need to create the user password hashes in the same manner that Solr expects them. The default `security.json` will define an `admin` user, a general purpose read-only `solr` user, and special `k8s-oper` user that *only* has read access to the API endpoints needed to operator in Kubernetes, such as the `/admin/info/system` endpoint used by probes. The password for each user can be retrieved using `kubectl` using:
```
kubectl get secret dev-solrcloud-basic-auth -o jsonpath='{.data.k8s-oper}' | base64 --decode
```

Also, I'm not sure if the operator should watch the `security.json` ConfigMap for updates (as we do with other user-provided ConfigMaps) and try to apply the updated JSON? There's an API for updating the security config and the ref guide suggests using that so maybe the operator just bootstraps the file and then it's up to users to maintain via the API? Trying to accommodate user updates via the API and/or ConfigMap could get messy.

For the probes, K8s allows specifying a header for HTTP endpoints, but then our password will live in the STS config. I think it would be better to set an env var from the secret and then run a command for the probe using this env var. The SolrCLI command-line tool provides an `api` action that we can use for the probes:
```
java -Dbasicauth=k8s-oper:****** \
  -Dsolr.httpclient.builder.factory=org.apache.solr.client.solrj.impl.PreemptiveBasicAuthClientBuilderFactory \
  -Dsolr.install.dir="/opt/solr" \
  -Dlog4j.configurationFile="/opt/solr/server/resources/log4j2-console.xml" \
  -classpath "/opt/solr/server/solr-webapp/webapp/WEB-INF/lib/*:/opt/solr/server/lib/ext/*:/opt/solr/server/lib/*" \
  org.apache.solr.util.SolrCLI api -get http://localhost:8983/solr/admin/info/system
```
Unfortunately, `bin/solr` does not expose the `api` action :-( So we'll need to take some liberties on how the probe command is constructed, such as assuming Solr is installed at `/opt/solr`

Still testing / tweaking the default `security.json` but I imagine it would be something like this:
```
{
      "authentication":{
        "blockUnknown": true,
        "class":"solr.BasicAuthPlugin",
        "credentials": {
          "admin":  "...",
          "k8s-oper":     "...",
          "solr":   "..."
        },
        "realm":"Solr Basic Auth",
        "forwardCredentials": false
      },
      "authorization": {
         "class": "solr.RuleBasedAuthorizationPlugin",
         "user-role": {
            "admin": ["admin", "probe"],
            "k8s-oper": ["probe"],
            "solr": ["users", "probe"]
         },
         "permissions": [
           { "name": "k8s-probes", "role":"probe", "collection": null, "path":"/admin/info/system" },
           { "name": "k8s-status", "role":"probe", "collection": null, "path":"/admin/collections" },
           { "name": "all", "role":["admin","users"] },
           { "name": "read", "role":["admin","users"] },
           { "name": "update", "role":["admin"] },
           { "name": "security-read", "role": "admin"},
           { "name": "security-edit", "role": "admin"}
         ]
      }
    }
```

### PR DESCRIPTION
**Pull request by thelabdude**
This PR fixes #218 

The changes in this PR allow users to bootstrap a `security.json` for a SolrCloud cluster with basic auth enabled.
```
spec:
  ...
  solrSecurity:
    authenticationType: Basic
```
A word about the probes ... we recommend configuring Solr to allow un-authenticated access over HTTP to the probe endpoint(s) and the bootstrapped `security.json` does this for you automatically. However, if you want secure the probe endpoints, then you need to set `probesRequireAuth: true`:
```
spec:
  ...
  solrSecurity:
    authenticationType: Basic
    probesRequireAuth: true
```
When auth is enabled, the liveness & readiness probes are configured to execute a command instead of using HTTP. Of course you can supply the credentials in an `Authorization` header for the HTTP probes, but that would require a restart of all pods if the password changes. With a command, we can load the credentials from file that comes from a secret. K8s will update the secret file when the secret changes automatically. You can see this work by changing the password in Solr first (using the security API), e.g.
```
curl --user $CREDS http://localhost:8983/solr/admin/authentication \
   -H 'Content-type:application/json' -d '{"set-user": {"k8s-oper":"Test1234"}}'
```
If you wait a bit, then probes will start to fail b/c the password has changed on the Solr side but is not yet reflected in the operator / k8s layer. Edit the secret to update the password for the `k8s-oper` user and you'll see probes begin to work again after a brief delay. If you look behind the scenes, you'll notice the command used for probes invokes the `SolrCLI` utility:
```
java  org.apache.solr.util.SolrCLI api -get http://localhost:8983/solr/admin/info/system
```
Unfortunately, the `bin/solr` command does not expose the `api` action we need, so we have to invoke `SolrCLI` directly. 

See user docs included in this PR for additional info ...

**TODO:**
- [x] User docs
- [x] Auth for Backup / Restore API
- [x] Prometheus Exporter config
- [x] Mount secret used by probe command as file (instead of env var) so that K8s automatically updates it when the secret changes

### CODE REVIEW

File: controllers/util/solr_util.go
Comment (thelabdude) [inline controllers/util/solr_util.go:1010]: haha ... don't laugh at this ;-) it was the easiest way to check if the znode data is empty "" or {} using /bin/sh

File: controllers/util/solr_util.go
Comment (thelabdude) [inline controllers/util/solr_util.go:1351]: See Solr's Sha256AuthenticationProvider: https://github.com/apache/lucene-solr/blob/master/solr/core/src/java/org/apache/solr/security/Sha256AuthenticationProvider.java#L114

**Comment by HoustonPutman** [comment]
A few thoughts:

- I completely agree that only files used in the filesystem in Solr (solr.xml, log4j, prometheus exporter config) should be stored in configMaps. Those that are merely meant to bootstrap Zookeeper, should not be stored in configMaps.
- I like the initAnonymousEndpoints, but just so that I'm clear, that's just for the bootstrapped security.json right? Even if the user changes that list after the cluster is created, it won't be reflected in the security.json? If so maybe we leave it up to the users to use the security APIs to manage that. Not really sure yet either way.
- For the HTTP Probes, would it not be easy to merely remove the auth from those initially? Or I guess we can't tell whether the user has removed auth for it or not, so we need to provide it by default.

File: controllers/solrcloud_controller.go
Comment (HoustonPutman) [inline controllers/solrcloud_controller.go:301]: shouldn't you only set this if probes should use the auth data?

Also maybe we should rename this from configMapInfo to something else...

File: controllers/solrprometheusexporter_controller.go
Comment (HoustonPutman) [inline controllers/solrprometheusexporter_controller.go:306]: Honestly think at this point we should bake something into ControllerRuntime to automate this. We use it so often.

(Nothing to do with this PR, just an observation)

File: controllers/solrprometheusexporter_controller.go
Comment (HoustonPutman) [inline controllers/solrprometheusexporter_controller.go:220]: We might have to look in two places here. It could be in `data` or `stringData`.

https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret

File: api/v1beta1/solrcloud_types.go
Comment (HoustonPutman) [inline api/v1beta1/solrcloud_types.go:1063]: I am not familiar with how other CRDs handle this, but should the "key" have to be the username, or should it be passed in separately?

Like `basicAuthUsername` and `basicAuthPasswordSecret`. 

[Looking at the kubernetes docs](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret), it looks like the standard for Basic Auth secrets is to store "username": <> and "password": <>.

File: api/v1beta1/solrprometheusexporter_types.go
Comment (HoustonPutman) [inline api/v1beta1/solrprometheusexporter_types.go:98]: Same comment about basicAuth secrets standards.

File: controllers/solrcloud_controller.go
Comment (HoustonPutman) [inline controllers/solrcloud_controller.go:274]: Could be in Data or StringData

File: controllers/util/solr_util.go
Comment (HoustonPutman) [inline controllers/util/solr_util.go:1245]: We might want to use `Type: "kubernetes.io/basic-auth"`?

File: controllers/solrcloud_controller.go
Comment (HoustonPutman) [inline controllers/solrcloud_controller.go:305]: does the Data need to be decoded from base64?

**Review by HoustonPutman — COMMENTED** [review]
Reviewed everything except for solr_util.go.

Main question I have for now is the format of basic auth secrets that we should accept. It seems like [Kube has a standard](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret).

File: api/v1beta1/solrcloud_types.go
Comment (thelabdude) [reply api/v1beta1/solrcloud_types.go:1063]: We're storing creds for multiple users: admin, solr, and k8s-oper as well as the security.json (so we don't have to recompute it for every reconcile loop). So our secret is not a K8s basic auth secret

File: api/v1beta1/solrcloud_types.go
Comment (thelabdude) [reply api/v1beta1/solrcloud_types.go]: ok, I removed it

File: controllers/solrprometheusexporter_controller.go
Comment (thelabdude) [reply controllers/solrprometheusexporter_controller.go:220]: afaik, `stringData` is a convenience for YAML and secrets always store all data in `data` but I'll verify if they supply via `stringData` that we can read it from `data`

Comment (HoustonPutman) [reply controllers/solrprometheusexporter_controller.go:220]: Ahhh very cool, did not know that!

Comment (thelabdude) [reply controllers/solrprometheusexporter_controller.go:220]: Confirmed that only `data` is stored ... here's an example:
```
apiVersion: v1
stringData:
  k8s-oper: Test1234
kind: Secret
metadata:
  name: user-provided-basic-auth
type: Opaque
```
and `kubectl get secret user-provided-basic-auth -o yaml`  gives:
```
apiVersion: v1
data:
  k8s-oper: VGVzdDEyMzQ=
kind: Secret
metadata:
  name: user-provided-basic-auth
type: Opaque
```
and in the logs, we have:
```
2021-02-19T19:38:56.486Z	INFO	controllers.SolrCloud	For user-supplied secret, password is:	{"namespace": "dev", "solrCloud": "dev", "password": "Test1234"}
```
_Note: I just added that log message to test this scenario ... it has since been removed, i.e. we're not logging passwords_

File: api/v1beta1/solrcloud_types.go
Comment (thelabdude) [reply api/v1beta1/solrcloud_types.go:1083]: One idea I liking more and more is to have 2 secrets, one that is a `"kubernetes.io/basic-auth"` that holds the creds for the `k8s-oper` user and another `Opaque` that holds the bootstrapped `admin` user and the `security.json`. We can name the latter `<cloud>-solrcloud-security-bootstrap` (or similar) which makes it clear it's only for bootstrapping security and the former holds the creds for the user needed by the operator.

**Comment by sylus** [comment]
Hi there I ended up testing this P.R. and it works great in a non Istio enabled namespace.

However the problem is with the solr-zk-setup init container when is in an istio enabled namespace as there is no network connectivity.

The Istio CNI plugin may cause networking connectivity problems for any application initContainers. When using Istio CNI kubelet starts an injected pod with the following steps:

1. The Istio CNI plugin sets up traffic redirection to the Istio sidecar proxy within the pod
2. All init containers execute and complete successfully
3. The Istio sidecar proxy starts in the pod along with the pods other containers

Init containers execute before the sidecar proxy starts, which can results in traffic loss during their exection. Some workarounds include:

a) Set the traffic.sidecar.istio.io/excludeOutboundIPRanges
b) Set the traffic.sidecar.istio.io/excludeOutboundPorts
c) Running initContainers as user 1337 since that bypasses Istio

I am waiting for zookeeper-operator to allow for custom annotations to be passed then this should work.

**Comment by thelabdude** [comment]
@sylus thanks for testing, really appreciate it ... I'll try to work in a solution using your work-arounds

**Comment by sylus** [comment]
I also had do this:

https://github.com/apache/lucene-solr-operator/issues/223

To get the quorum property but Solr then worked talking to ZooKeeper when was added (except for init container)

And thanks so much for the awesome work it was pretty cool to see the basic auth setup just work and not have to do anything :)

File: controllers/solrcloud_controller.go
Comment (thelabdude) [reply controllers/solrcloud_controller.go]: I renamed it to `reconcileConfigInfo` since it holds additional config info computed during reconcile ... ;-)

**Comment by thelabdude** [comment]
@sylus Let's tackle the Istio related changes in a follow-on PR from this one, i.e. no work-arounds for that here ;-) Can help with the follow-on PR once this is merged though, should be close ...

**Comment by sylus** [comment]
Thanks @thelabdude that makes complete sense to me :D

**Comment by sylus** [comment]
Once pravega allows for custom annotations technically wouldn't need any changes to your P.R. so that might be the best route forward :D

File: docs/solr-cloud/solr-cloud-crd.md
Comment (HoustonPutman) [inline docs/solr-cloud/solr-cloud-crd.md:596]: I think this would fit better as a `###` section after `Option 2: ...`

File: docs/solr-cloud/solr-cloud-crd.md
Comment (HoustonPutman) [inline docs/solr-cloud/solr-cloud-crd.md:773]: I would re-iterate here the endpoints needed for the prometheus exporter auth user, when moving it to the prometheus exporter docs.

File: docs/solr-cloud/solr-cloud-crd.md
Comment (HoustonPutman) [inline docs/solr-cloud/solr-cloud-crd.md:693]: We probably want to add `/admin/info/health` for the future as well. (this would affect multiple places in the PR)

File: controllers/util/solr_util.go
Comment (HoustonPutman) [inline controllers/util/solr_util.go:1243]: I think these two are validated by the API server, but not a bad thing to have it here as well.

**Review by HoustonPutman — COMMENTED** [review]
Looks really good! Just a few small things to finish this up

File: controllers/util/solr_util.go
Comment (thelabdude) [reply controllers/util/solr_util.go:1243]: Sadly not the case ... I tried creating a basic auth secret w/o the username and K8s happily obliged me :(

File: controllers/solrcloud_controller.go
Comment (thelabdude) [reply controllers/solrcloud_controller.go:313]: ok good idea, let me test that ... I think it would still trigger a full rolling restart of the STS (b/c the initContainer def changes) but you're right we should fail if they want to delete it

File: api/v1beta1/solrcloud_types.go
Comment (thelabdude) [reply api/v1beta1/solrcloud_types.go:1054]: I considered that and chose to make it required by the user to set to `Basic` but I don't have strong feelings about enforcing that, will make optional with default `Basic`

File: docs/solr-cloud/solr-cloud-crd.md
Comment (thelabdude) [reply docs/solr-cloud/solr-cloud-crd.md:693]: We do you think we should do about the `customPodOptions` probes and auth? Currently, if the user supplies a custom  probe config with a different path and has `probesRequireAuth`, the code is not picking up the custom path. If `probesRequireAuth` is `false`, then the custom path does get added to the authz rules correctly. This is a bit trappy but I don't know if the operator should try to swap out their custom probe config with a command either? So perhaps we just doc this limitation for now?

File: controllers/util/solr_util.go
Comment (HoustonPutman) [reply controllers/util/solr_util.go:1243]: Very strange. Well good to have the check then!

File: controllers/solrcloud_controller.go
Comment (HoustonPutman) [reply controllers/solrcloud_controller.go:313]: definitely would cause a full rolling restart, but at least their cloud would be able to reconcile 😅

File: api/v1beta1/solrcloud_types.go
Comment (HoustonPutman) [reply api/v1beta1/solrcloud_types.go:1054]: Don't have strong feelings either. Up to you!

File: docs/solr-cloud/solr-cloud-crd.md
Comment (thelabdude) [reply docs/solr-cloud/solr-cloud-crd.md:693]: If we go with the punt on `probesRequireAuth` and custom probe paths approach, I think we should also error out during reconcile so it's clear we don't like their config, i.e. something along the lines of:
```
Found custom probe HTTP paths and probesRequireAuth=true, this configuration is not currently supported. Please switch to probesRequireAuth=false when using custom probe HTTP paths.
```

Comment (thelabdude) [reply docs/solr-cloud/solr-cloud-crd.md:707]: I've included `/admin/info/health` as a default in the rules for now

Comment (thelabdude) [reply docs/solr-cloud/solr-cloud-crd.md:728]: Re-worked this a bit and improved the docs in this area. Added a check for how custom probe paths are handled to one of the existing unit tests

**Review by HoustonPutman — APPROVED** [review]
This is good to go. Thoroughly tested it locally, and couldn't find any issues.

Awesome feature, and huge win for the community!

Thanks for the great work @thelabdude!!

**Pull request merged**
